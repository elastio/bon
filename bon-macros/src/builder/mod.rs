mod builder_gen;
mod params;

pub(crate) mod item_impl;

mod item_func;
mod item_struct;

use crate::normalization::{ExpandCfg, ExpansionOutput};
use crate::util;
use crate::util::prelude::*;
use darling::FromMeta;
use quote::quote;
use syn::parse::Parser;

pub(crate) fn generate_from_derive(item: TokenStream2) -> TokenStream2 {
    try_generate_from_derive(item).unwrap_or_else(Error::write_errors)
}

fn try_generate_from_derive(item: TokenStream2) -> Result<TokenStream2> {
    match syn::parse2(item)? {
        syn::Item::Struct(item_struct) => item_struct::generate(item_struct),
        _ => bail!(
            &Span::call_site(),
            "only `struct` items are supported by the `#[derive(bon::Builder)]` attribute"
        ),
    }
}

pub(crate) fn generate_from_attr(params: TokenStream2, item: TokenStream2) -> TokenStream2 {
    try_generate_from_attr(params.clone(), item.clone()).unwrap_or_else(|err| {
        [
            generate_completion_triggers(params),
            crate::error::error_into_token_stream(err, item),
        ]
        .concat()
    })
}

fn try_generate_from_attr(params: TokenStream2, item: TokenStream2) -> Result<TokenStream2> {
    let item: syn::Item = syn::parse2(item)?;

    if let syn::Item::Struct(item_struct) = item {
        return Ok(quote! {
            // Triggers a deprecation warning if the user is using the old attribute
            // syntax on the structs instead of the derive syntax.
            use ::bon::private::deprecations::builder_attribute_on_a_struct as _;

            #[derive(::bon::Builder)]
            #[builder(#params)]
            #item_struct
        });
    }

    let macro_path = syn::parse_quote!(::bon::builder);

    let ctx = ExpandCfg {
        macro_path,
        params,
        item,
    };

    let (params, item) = match ctx.expand_cfg()? {
        ExpansionOutput::Expanded { params, item } => (params, item),
        ExpansionOutput::Recurse(output) => return Ok(output),
    };

    let nested_meta = &darling::ast::NestedMeta::parse_meta_list(params.clone())?;

    let main_output = match item {
        syn::Item::Fn(item_fn) => item_func::generate(FromMeta::from_list(nested_meta)?, item_fn)?,
        _ => bail!(
            &Span::call_site(),
            "only `fn` items are supported by the `#[bon::builder]` attribute"
        ),
    };

    let output = [generate_completion_triggers(params), main_output].concat();

    Ok(output)
}

fn generate_completion_triggers(params: TokenStream2) -> TokenStream2 {
    let meta = util::ide::parse_comma_separated_meta
        .parse2(params)
        .unwrap_or_default();

    util::ide::generate_completion_triggers(meta)
}
