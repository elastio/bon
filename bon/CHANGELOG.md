# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.1](https://github.com/elastio/bon/compare/bon-v0.1.0...bon-v0.1.1) - 2024-07-23

### Other
- Upd docs
- More improvements, deploy-website workflow
- Docs for alternatives plus `#[builder(name = ...)]` for members
- Rename `field -> member`
- Add more docs
- Add doctests, more docs
- More docs
- More docs and better decisions
- Fix builder attrs not being removed correctly
- Next iteration
- Test struct input
- Start testing struct input
- Refactor in preparation for struct input for macro
- Make `#[default]` compatible with `Option`
- Add `default` and `into` attrs
- Make bool optional, strip `_` prefix
- Add special casing for `Option` setters
- Better doc comments
- Improve nested refs handling
- Cleanup
- Fix a bunch of papercuts with codegen
- Finish implementing support for associated functions
- Start adding support for associated functions
- Add some docs
- Rebranding buildy -> prox (buildy crate name is already taken on crates.io)
