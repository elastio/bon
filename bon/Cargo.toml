[package]
name    = "bon"
version = "2.3.0"

description = "Next-gen compile-time-checked builder generator, named function's arguments, and more!"

categories = [
  "rust-patterns",
  "data-structures",
  "asynchronous",
  "no-std",
  "no-std::no-alloc",
]
keywords = ["builder", "macro", "derive", "constructor", "setter"]

edition    = "2021"
homepage   = "https://bon-rs.com/"
license    = "MIT OR Apache-2.0"
repository = "https://github.com/elastio/bon"

# This MSRV was chosen because Rust supports mixing generic type and const
# parameters only starting with this version. We require this feature for the
# cases when the builder is generated for a function or struct that uses const
# generics. The generated builder **always** appends a generic type parameter
# for the type state at the end of the Builder generics list. So for functions
# or structs that use const generics the generated builder will have a generic
# type parameter after the const generic parameter, which is only supported
# starting from Rust 1.59.0.
rust-version = "1.59.0"

[package.metadata.docs.rs]
all-features = true

# Generate clickable links in the source code view in the docs
rustdoc-args = ["--generate-link-to-definition"]

# We don't need the docs to be built for every first-tier target.
# This crate is not platform-specific.
targets = ["x86_64-unknown-linux-gnu"]

[lints]
workspace = true

[dependencies]
# The version of the macro crate is pinned to a specific one because the code
# generated by the macros uses private APIs from the runtime crate that are not
# guarded by semver.
bon-macros  = { path = "../bon-macros", version = "=2.3.0" }
rustversion = "1"

[dev-dependencies]
# Using a bit older version that supports our MSRV
expect-test = "1.4.1"

# Using a bit older version that supports our MSRV
tokio = { version = "1.29.1", features = ["macros", "rt-multi-thread"] }

# Using a bit older version that supports our MSRV
trybuild = "1.0.89"

[features]
alloc   = []
default = ["std"]
std     = ["alloc"]

# Opts in to the higher MSRV 1.79.0. In this version, Rust stabilized the syntax
# for bounds in associated type position which can be used to make bounds on generic
# associated types implied. See the release announcement for more:
# https://blog.rust-lang.org/2024/06/13/Rust-1.79.0.htmlbounds-in-associated-type-position
#
# This feature is useful for the trait `IsComplete` generated by the builder macros.
# When this feature is enabled, the builder macros use the new syntax for bounds in
# associated type position, which enables implied `IsSet` bounds for the type state
# of required members.
#
# To understand how this can be used consider the following example:
#
# ```rust
# #[derive(bon::Builder)]
# struct Example {
#     a: u32,
#     b: Option<u32>,
# }
#
# use example_builder::{IsUnset, IsComplete};
#
# impl<State: example_builder::State> ExampleBuilder<State> {
#     fn build_with_default_b(self) -> Example
#     where
#         State: IsComplete,
#         State::B: IsUnset,
#     {
#         self.b(42).build()
#     }
# }
# ```
#
# This code wouldn't compile without this feature enabled, because `State: IsComplete`
# wouldn't automatically imply `State::A: IsSet`, so the builder type state returned
# after `self.b()` doesn't imply that the member `a` is set, and thus `build()`
# can't be called.
#
# Huge thanks to @harudagondi for suggesting the name of this cargo feature!
implied-bounds = ["bon-macros/implied-bounds"]

# üî¨ Experimental! There may be breaking changes to this feature between *minor* releases,
# however, compatibility within patch releases is guaranteed though.
#
# This feature enables the #[builder(overwritable)] attribute that can be used to
# allow overwriting already set fields in the builder. Without this attribute it's
# not possible to set the same member twice in the builder, a compile error would be
# generated.
#
# See more info at https://bon-rs.com/reference/builder/top-level/overwritable.
#
# We are considering stabilizing this attribute if you have a use for it. Please leave
# a üëç reaction under the issue https://github.com/elastio/bon/issues/149 if you need
# this attribute. It would also be cool if you could leave a comment under that issue
# describing your use case for it.
experimental-overwritable = ["bon-macros/experimental-overwritable"]
