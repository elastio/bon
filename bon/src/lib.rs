#![doc = include_str!("../README.md")]

pub use bon_macros::*;

/// Symbols used by macros. They are not stable and are considered an implementation detail.
#[doc(hidden)]
pub mod private;

/// Same as [`std::vec!`] but converts each element with [`Into::into()`].
///
/// **WARNING:** it's not recommended to import this macro into scope. Reference it
/// using the full path (`bon::vec![]`) to avoid confusion with the [`std::vec!`] macro.
///
/// A good example of the use case for this macro is when you want to create a
/// [`Vec<String>`] where part of the items are hardcoded string literals of type
/// `&str` and the other part is made of dynamic [`String`] values.
///
/// ```
/// fn convert_media(input_extension: &str, output_extension: &str) -> std::io::Result<()> {
///     let ffmpeg_args: Vec<String> = bon::vec![
///         "-i",
///         format!("input.{input_extension}"),
///         "-y",
///         format!("output.{output_extension}"),
///     ];
///
///     std::process::Command::new("ffmpeg").args(ffmpeg_args).output()?;
///
///     Ok(())
/// }
/// ```
///
/// This macro doesn't support `vec![expr; N]` syntax, since it's simpler to
/// just write `vec![expr.into(); N]` using [`std::vec!`] instead.
#[macro_export]
macro_rules! vec {
    () => (::std::vec::Vec::new());
    ($($item:expr),+ $(,)?) => (::std::vec![$(::core::convert::Into::into($item)),+ ]);
}

/// Creates a fixed-size array literal where each element is converted with [`Into::into()`]
/// into the target type.
///
/// You'll probably need a hint for the target type of items in the array if the
/// compiler can't infer it from its usage.
///
/// This is similar in spirit to the [`bon::vec!`] macro, but it's for arrays.
/// See [`bon::vec!`] docs for details.
///
/// Same example as in [`bon::vec!`], but using this macro. It works with array
/// as well because [`Command::args`] accepts any value that implements [`IntoIterator`]:
///
/// ```
/// fn convert_media(input_extension: &str, output_extension: &str) -> std::io::Result<()> {
///     let ffmpeg_args: [String; 4] = bon::arr![
///         "-i",
///         format!("input.{input_extension}"),
///         "-y",
///         format!("output.{output_extension}"),
///     ];
///
///     std::process::Command::new("ffmpeg").args(ffmpeg_args).output()?;
///
///     Ok(())
/// }
/// ```
///
/// This macro doesn't support `[expr; N]` syntax, since it's simpler to
/// just write `[expr.into(); N]` instead.
///
/// [`Command::args`]: std::process::Command::args
/// [`bon::vec!`]: crate::vec
#[macro_export]
macro_rules! arr {
    () => ([]);
    ($($item:expr),+ $(,)?) => ([$(::core::convert::Into::into($item)),+]);
}

#[cfg(test)]
mod tests {
    #[test]
    fn arr_smoke() {
        let actual: [String; 3] = crate::arr!["foo", "bar", "baz"];
        assert_eq!(actual, ["foo", "bar", "baz"]);

        let actual: [String; 0] = crate::arr![];
        assert!(actual.is_empty());
    }

    #[test]
    fn vec_smoke() {
        let actual: Vec<String> = crate::vec!["foo", "bar", "baz"];
        assert_eq!(actual, ["foo", "bar", "baz"]);

        let actual: Vec<String> = crate::vec![];
        assert!(actual.is_empty());
    }
}
