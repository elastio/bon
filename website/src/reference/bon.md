# `#[bon]`

This is a companion macro for [`builder`]. You should place it on top of the `impl` block
where you want to define methods with the [`builder`] macro. It provides
the necessary context to the [`builder`] macros on top of the functions
inside of the `impl` block. You'll get compile errors without that context.

For the examples of the usage of this macro and the reason why it's needed, see this paragraph in [the overview](../guide/overview#builder-for-an-associated-method).

[`builder`]: ./builder
