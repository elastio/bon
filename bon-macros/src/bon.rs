use crate::builder;
use crate::normalization::{ExpandCfg, ExpansionOutput};
use crate::util::prelude::*;

pub(crate) fn generate(params: TokenStream2, item: TokenStream2) -> TokenStream2 {
    try_generate(params, item.clone())
        .unwrap_or_else(|err| crate::error::error_into_token_stream(err, item))
}

pub(crate) fn try_generate(params: TokenStream2, item: TokenStream2) -> Result<TokenStream2> {
    let item: syn::Item = syn::parse2(item)?;
    let macro_path = syn::parse_quote!(::bon::bon);

    let ctx = ExpandCfg {
        macro_path,
        params,
        item,
    };

    let (_params, item) = match ctx.expand_cfg()? {
        ExpansionOutput::Expanded { params, item } => (params, item),
        ExpansionOutput::Recurse(output) => return Ok(output),
    };

    match item {
        syn::Item::Impl(item_impl) => builder::item_impl::generate(item_impl),
        _ => bail!(
            &item,
            "`#[bon]` attribute is expected to be placed on an `impl` block \
             but it was placed on other syntax instead"
        ),
    }
}
