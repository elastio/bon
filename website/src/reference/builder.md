---
outline: deep
---

# `#[derive(Builder)]` / `#[builder]`

Generate builders from structs via `#[derive(Builder)]`, free functions via `#[builder]`, and associated methods via `#[bon]` + `#[builder]`. They all use the same attributes syntax.

## Top-Level Attributes

These attributes are placed on top of a `struct` or `fn` declaration.

| Attribute                                          | Short description                                                                                    |
| -------------------------------------------------- | ---------------------------------------------------------------------------------------------------- |
| [`builder_type`](./builder/top-level/builder_type) | Overrides name, visibility and docs for the builder struct                                           |
| [`crate`](./builder/top-level/crate)               | Overrides path to `bon` crate referenced in the generated code                                       |
| [`derive`](./builder/top-level/derive)             | Generates additional derives for the builder struct itself                                           |
| [`finish_fn`](./builder/top-level/finish_fn)       | Overrides name, visibility and docs for the finishing function                                       |
| [`on`](./builder/top-level/on)                     | Applies member attributes to all members matching a type pattern                                     |
| [`start_fn`](./builder/top-level/start_fn)         | Overrides name, visibility and docs for the starting function                                        |
| [`state_mod`](./builder/top-level/state_mod)       | Overrides name, visibility and docs for the builder's [typestate API](../guide/typestate-api) module |

## Member Attributes

These attributes are placed on a `struct` field or `fn` argument.

| Attribute                                          | Short description                                                |
| -------------------------------------------------- | ---------------------------------------------------------------- |
| [`default`](./builder/member/default)              | Makes the member optional with a default value                   |
| [`finish_fn`](./builder/member/finish_fn)          | Makes the member a positional argument on the finishing function |
| [`into`](./builder/member/into)                    | Changes the signature of the setters to accept `impl Into<T>`    |
| [`name`](./builder/member/name)                    | Overrides the name of the member used in the builder's API       |
| [`overwritable` ðŸ”¬](./builder/member/overwritable) | Allows calling setters for the same member repeatedly            |
| [`setters`](./builder/member/setters)              | Overrides name, visibility and docs for setters                  |
| [`skip`](./builder/member/skip)                    | Skips generating setters for the member                          |
| [`start_fn`](./builder/member/start_fn)            | Makes the member a positional argument on the starting function  |
| [`transparent`](./builder/member/transparent)      | Disables `Option<T>` special handling, makes the member required |
| [`with`](./builder/member/with)                    | Overrides setters' signature and applies a custom conversion     |

## Examples

:::code-group

```rust [Struct]
use bon::Builder;

#[derive(Builder)]
#[builder(finish_fn = finish)] // <-- this is a top-level attribute // [!code highlight]
struct Example {
    #[builder(default)] // <-- this is a member attribute // [!code highlight]
    field: u32
}
```

```rust [Function]
use bon::builder;

#[builder(finish_fn = finish)] // <-- this is a top-level attribute // [!code highlight]
fn example(
    #[builder(default)] // <-- this is a member attribute // [!code highlight]
    arg: u32
) { }
```

```rust [Method]
use bon::bon;

struct Example;

#[bon]
impl Example {
    #[builder(finish_fn = finish)] // <-- this is a top-level attribute // [!code highlight]
    fn example(
        #[builder(default)]  // <-- this is a member attribute // [!code highlight]
        arg: u32
    ) { }
}
```
