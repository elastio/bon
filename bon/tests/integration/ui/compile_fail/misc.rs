use bon::builder;
use std::collections::{BTreeMap, BTreeSet};

#[builder]
struct TupleStruct(u32, u32);

#[builder]
fn destructuring((x, y): (u32, u32)) {
    let _ = x;
    let _ = y;
}

#[builder]
fn explicit_into_equals_true(#[builder(into = true)] _x: u32) {}

#[builder]
fn unnecessary_into_override_true(#[builder(into)] _x: String) {}

#[builder]
fn unnecessary_into_override_false(#[builder(into = false)] _x: u32) {}

#[builder(start_fn())]
struct EmptyStartFn {}

fn main() {
    let map: BTreeMap<String, String> = bon::map! {
        "Hello": "Blackjack",
        "Hello": "Littlepip",
    };

    let set: BTreeSet<String> = bon::set!["mintals", "guns", "mintals", "roses"];
}
