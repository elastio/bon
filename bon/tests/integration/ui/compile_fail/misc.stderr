error: duplicate key in the map
 --> tests/integration/ui/compile_fail/misc.rs:6:9
  |
6 |         "Hello": "Blackjack",
  |         ^^^^^^^

error: duplicate key in the map
 --> tests/integration/ui/compile_fail/misc.rs:7:9
  |
7 |         "Hello": "Littlepip",
  |         ^^^^^^^

error: duplicate value in the set
  --> tests/integration/ui/compile_fail/misc.rs:10:43
   |
10 |     let set: BTreeSet<String> = bon::set!["mintals", "guns", "mintals", "roses"];
   |                                           ^^^^^^^^^

error: duplicate value in the set
  --> tests/integration/ui/compile_fail/misc.rs:10:62
   |
10 |     let set: BTreeSet<String> = bon::set!["mintals", "guns", "mintals", "roses"];
   |                                                              ^^^^^^^^^

error: Only structs with named fields are supported
  --> tests/integration/ui/compile_fail/misc.rs:26:1
   |
26 | struct TupleStruct(u32, u32);
   | ^^^^^^

error: use a simple `identifier: type` syntax for the function argument; destructuring patterns in arguments aren't supported by the `#[builder]`
  --> tests/integration/ui/compile_fail/misc.rs:29:18
   |
29 | fn destructuring((x, y): (u32, u32)) {
   |                  ^^^^^^

error: Unexpected type `bool`
  --> tests/integration/ui/compile_fail/misc.rs:35:44
   |
35 | fn unnecessary_into_false(#[builder(into = false)] _x: u32) {}
   |                                            ^^^^^

error: this `#[builder(into)]` attribute is redundant, because `into` is already implied for this member via the `#[builder(on(...))]` at the top of the function
  --> tests/integration/ui/compile_fail/misc.rs:38:31
   |
38 | fn unnecessary_into(#[builder(into)] _x: String) {}
   |                               ^^^^

error: This syntax is not supported in type patterns yet. If you have a use case for this, please open an issue at https://github.com/elastio/bon/issues.
  --> tests/integration/ui/compile_fail/misc.rs:40:15
   |
40 | #[builder(on(&dyn std::fmt::Debug, into))]
   |               ^^^

error: nested attributes are not allowed in the type pattern of #[builder(on(type_pattern, ...))]
  --> tests/integration/ui/compile_fail/misc.rs:43:17
   |
43 | #[builder(on(fn(#[attr] a: u32), into))]
   |                 ^

error: Expected an attribute of form `on(type_pattern, ...)`
  --> tests/integration/ui/compile_fail/misc.rs:46:11
   |
46 | #[builder(on)]
   |           ^^

error: unexpected end of input, expected one of: `for`, parentheses, `fn`, `unsafe`, `extern`, identifier, `::`, `<`, `dyn`, square brackets, `*`, `&`, `!`, `impl`, `_`, lifetime
  --> tests/integration/ui/compile_fail/misc.rs:49:1
   |
49 | #[builder(on())]
   | ^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `builder` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected `,`
  --> tests/integration/ui/compile_fail/misc.rs:52:1
   |
52 | #[builder(on(_))]
   | ^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `builder` (in Nightly builds, run with -Z macro-backtrace for more info)

error: this #[builder(on(type_pattern, ...))] contains no options to override the default behavior for the selected setters like `into`, so it does nothing
  --> tests/integration/ui/compile_fail/misc.rs:55:1
   |
55 | #[builder(on(_,))]
   | ^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `builder` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected at least one parameter in parentheses
  --> tests/integration/ui/compile_fail/misc.rs:58:11
   |
58 | #[builder(start_fn())]
   |           ^^^^^^^^

error: `skip` attribute can't be specified with other attributes like `into` because there will be no setter generated for this member to configure
  --> tests/integration/ui/compile_fail/misc.rs:63:21
   |
63 |     #[builder(skip, into)]
   |                     ^^^^

error: `skip` attribute can't be specified with other attributes like `name` because there will be no setter generated for this member to configure
  --> tests/integration/ui/compile_fail/misc.rs:69:28
   |
69 |     #[builder(skip, name = bar)]
   |                            ^^^

error: `skip` attribute can't be specified with other attributes like `default` because there will be no setter generated for this member to configure. If you wanted to specify a value for the member, then use the following syntax instead `#[builder(skip = value)]`
  --> tests/integration/ui/compile_fail/misc.rs:75:21
   |
75 |     #[builder(skip, default = 42)]
   |                     ^^^^^^^

error: `skip` attribute is not supported on function arguments. Use a local variable instead.
  --> tests/integration/ui/compile_fail/misc.rs:81:15
   |
81 |     #[builder(skip)] _x: u32,
   |               ^^^^

error: Only structs with named fields are supported
  --> tests/integration/ui/compile_fail/misc.rs:88:1
   |
88 | struct TupleStructsAreUnsupported(u32, u32);
   | ^^^^^^

error: The attribute is expected to be placed only on an `fn` item or a `struct` declaration
  --> tests/integration/ui/compile_fail/misc.rs:91:1
   |
91 | enum EnumsAreUnsupported {}
   | ^^^^

error: use a simple `identifier: type` syntax for the function argument; destructuring patterns in arguments aren't supported by the `#[builder]`
  --> tests/integration/ui/compile_fail/misc.rs:94:39
   |
94 | fn destructuring_in_fn_is_unsupported((_, _): (u32, u32)) {}
   |                                       ^^^^^^

error[E0599]: no method named `x` found for struct `SkipGeneratesNoSetterBuilder` in the current scope
  --> tests/integration/ui/compile_fail/misc.rs:21:38
   |
12 |     #[builder]
   |     ---------- method `x` not found for this struct
...
21 |     SkipGeneratesNoSetter::builder().x(42).build();
   |                                      ^ method not found in `SkipGeneratesNoSetterBuilder<()>`

error[E0599]: no method named `y` found for struct `SkipGeneratesNoSetterBuilder` in the current scope
  --> tests/integration/ui/compile_fail/misc.rs:22:38
   |
12 |     #[builder]
   |     ---------- method `y` not found for this struct
...
22 |     SkipGeneratesNoSetter::builder().y(42).build();
   |                                      ^ method not found in `SkipGeneratesNoSetterBuilder<()>`
